Ignore unconfigurable properties in casc config

meeusen opened the pull request:
<!-- Comment:
A great PR typically begins with the line below.
-->

<!-- in case you work on a Jira issue, replace XXXXX with the numeric part of the issue ID you created in Jira -->
<!-- in case you work on github issue -->
Close #527 
<!-- in case this PR solves Github issue use close #### or closes, closed, fix, fixes, fixed, resolve, resolves, resolved -->

<!-- Comment:
If the issue is not fully described in Jira / Github, add more information here (justification, pull request links, etc.).

 * We do not require Jira / Github issues for minor improvements.
 * Bug fixes should have a Jira / Github issue to facilitate the backporting process.
 * Major new features should have a Jira / Github issue.
-->

### Explanation
This is a combination of #770 and #650, to remove reservedBy from UI configuration and to ignore reservedBy/reservedTimestamp/note from CasC configuration.

Users can reserve lockable resources and add notes. These reservations and notes currently disappear when reloading CasC configuration, or when Jenkins simply restarts. This pull request attempts to fix that.

The reservedTimestamp and note were already copied when reconfiguring lockable resources in the UI. This happened in LockableResourceManager.configure(). Move this part to the setDeclaredResources(), so it also runs when configuring from CasC config. 

_reservedBy_, _reservedTimestamp_ and _note_ are now ignored if they are defined in CasC. _reservedBy_ is removed from UI configuration.

### Testing done

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, **you must describe** the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->
Added test ConfigurationAsCodeTest#should_keep_reservations_after_reload. This test also fails when leaving out my changes.

Did some manual tests. CasC configuration with 
1. resource with filled in reservedBy/note fields
2. resource with empty reservedBy/note fields
3. resource without reservedBy/note fields

Reserved/reassigned/unreserved the resources, updated the notes. When reloading configuration, resources 1, 2 and 3 kept their reservation/note from before the reload. Also when making changes in the UI configuration, the reservations were kept.


### Proposed upgrade guidelines
_reservedBy_, _reservedTimestamp_ and _note_ are now ignored if they are defined in CasC.

### Submitter checklist

- [x] The Jira / Github issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/main/content/_data/changelogs/weekly.yml)).
  - The changelog generator for plugins uses the **pull request title as the changelog entry**.
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during the upgrade.
- [x] There is automated testing or an explanation that explains why this change has no tests.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There is at least one (1) approval for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the **pull request title** and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically. See also [release-drafter-labels](https://github.com/jenkinsci/.github/blob/ce466227c534c42820a597cb8e9cac2f2334920a/.github/release-drafter.yml#L9-L50).
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] java code changes are tested by automated test.

meeusen commented on src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java:
```diff
@@ -1046,30 +1055,13 @@ public void changeQueueOrder(final String queueId, final int newPosition) throws
-
             try (BulkChange bc = new BulkChange(this)) {
-                // reset resources to default which are not currently locked
-                this.resources.removeIf(resource -> !resource.isLocked());
```
I had to remove this, otherwise reservations were removed after reloading configuration. But I don't understand why I don't see this problem in #770, so I think I'm not fully understanding everything that's happening. But it works in my tests :).

meeusen replied:
Understood now. The reservedBy field was removed from the UI, so it is not part anymore of the declaredResources argument supplied to setDeclaredResources() when saving UI configuration. This line of code removed the resources from the current list, and that's the only place where note/reservedBy values are available.

manosnoam commented:
Thank you for this PR @meeusen 
Can you add a few screenshots of a reserved resource when locked, before and after Jenkins restart?
And how is it later released ?

mPokornyETM commented on src/test/java/org/jenkins/plugins/lockableresources/ConfigurationAsCodeTest.java:
```diff
@@ -32,26 +35,28 @@ void should_support_configuration_as_code(JenkinsConfiguredWithCodeRule r) {
         assertEquals("Label_A", declaredResource.getLabels());
-        assertEquals("Reserved_A", declaredResource.getReservedBy());
-        assertEquals("Note A", declaredResource.getNote());
+        // not supported in JCaC
```
In that case shal this works
`
   assertEquals("", declaredResource.getReservedBy());
   assertEquals("", declaredResource.getNote());
`

meeusen replied:
I've updated it with your suggestion.

mPokornyETM commented on src/test/java/org/jenkins/plugins/lockableresources/ConfigurationAsCodeTest.java:
```diff
@@ -32,26 +35,28 @@ void should_support_configuration_as_code(JenkinsConfiguredWithCodeRule r) {
         assertEquals("Label_A", resource.getLabels());
-        assertEquals("Reserved_A", resource.getReservedBy());
-        assertEquals("Note A", resource.getNote());
+        // not supported in JCaC
```
the same here, check for new "correct" value

mPokornyETM reviewed (approved):
You have 2 comments in test scripts. It is not mandatory to change it, but I will be happy.
Please provide some screenshots for changed UI interface.

meeusen commented:
- I've added your suggestions.
- Below you can find a screenshot
- I've compared the output of http://localhost:8090/jenkins/lockable-resources/api/json?pretty=true, with this pull request, and from the master branch, and it is the same.
- I did test that (see _Testing done_ in the pull request description). I've also added a new resource with note/reservedBy to the CasC yaml files in the automated testing, just to make sure that it still works.

I did notice that the reservedBy and note from CasC yaml are loaded, when starting Jenkins from scratch. Then there is no _org.jenkins.plugins.lockableresources.LockableResourcesManager.xml_ file yet, and there is no _previous_ instance to copy note/reservedBy from. I'll see if I can change it, e.g. when the resource does not exist in the _previous_ instance, to set reservedBy/note to empty.

Screenshot:
![2025-04-23-203832_1711x842_scrot](https://github.com/user-attachments/assets/33f7d152-e907-4683-a43f-996fa3e83397)

meeusen commented:
> I did notice that the reservedBy and note from CasC yaml are loaded, when starting Jenkins from scratch. Then there is no _org.jenkins.plugins.lockableresources.LockableResourcesManager.xml_ file yet, and there is no _previous_ instance to copy note/reservedBy from. I'll see if I can change it, e.g. when the resource does not exist in the _previous_ instance, to set reservedBy/note to empty.

I've added a commit to address this issue. reservedBy and note are now also ignored when starting Jenkins from scratch, or when loading CasC yaml with new resources.

meeusen commented:
> Thank you for this PR @meeusen Can you add a few screenshots of a reserved resource when locked, before and after Jenkins restart? And how is it later released ?

@manosnoam I'm not sure what you mean. A reserved resource is released when someone unreserves it. You can try the version from this pull request yourself, you can find the hpi file here: https://ci.jenkins.io/job/Plugins/job/lockable-resources-plugin/job/PR-771/.
